File: src/content/docs/topics/cs/leetcode-150/2d-dynamic-programming/115-distinct-subsequences.py
def num_distinct(s, t):
    pass  # TODO: implement

def _run_tests():
    assert num_distinct('rabbbit', 'rabbit') == 3
    assert num_distinct('babgbag', 'bag') == 5
    assert num_distinct('abc', '') == 1
    assert num_distinct('', 'a') == 0
    assert num_distinct('abc', 'abc') == 1
    assert num_distinct('aaa', 'b') == 0
    # --- large-input timing ---
    import time as _t
    _t0 = _t.perf_counter()
    num_distinct('a' * 200, 'a' * 100)
    _ms = (_t.perf_counter() - _t0) * 1000
    print(f'perf num_distinct("a"*200, "a"*100): {_ms:.1f}ms')
    print('all tests pass')

if __name__ == '__main__':
    _run_tests()

File: src/content/docs/topics/cs/coding-problems/2d-dynamic-programming/115-distinct-subsequences.swift
// LEETCODE_TYPE: Solution
func expectEqual<T: Equatable>(
    _ actual: T,
    _ expected: T,
    _ message: String = "",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard actual == expected else {
        let detail = message.isEmpty ? "values differ" : message
        fatalError("\(file):\(line): \(detail). Expected \(expected), got \(actual)")
    }
}

func expectTrue(
    _ condition: @autoclosure () -> Bool,
    _ message: String = "expected true",
    file: StaticString = #fileID,
    line: UInt = #line
) {
    guard condition() else {
        fatalError("\(file):\(line): \(message)")
    }
}

func reportSuccess() {
    print("All Swift tests passed")
}

final class Solution {
    func numDistinct(_ s: String, _ t: String) -> Int {
        // TODO: Implement
        fatalError("TODO: Implement")
    }
}

func runTests() {
    // TEST_VECTORS_BEGIN sha256:b52ea1d475aa012e676844982885b796ac2543f0c5aa4f51ef0905e594c44a4b
    expectEqual(Solution().numDistinct("rabbbit", "rabbit"), 3, "rabbbit-rabbit")
    expectEqual(Solution().numDistinct("babgbag", "bag"), 5, "babgbag-bag")
    expectEqual(Solution().numDistinct("abc", "abc"), 1, "same")
    expectEqual(Solution().numDistinct("abc", "abcd"), 0, "target-longer")
    expectEqual(Solution().numDistinct("abc", ""), 1, "empty-target")
    // EXCLUDED_VECTOR uppercase: ["ABC","A"] | Published inputs use English letters with documented case constraints.
    // EXCLUDED_VECTOR digit: ["a1","a"] | Published inputs contain letters only.
    // TEST_VECTORS_END
    reportSuccess()
}

runTests()
